automata/fa/dfa.py

```diff
@@ -581,6 +601,99 @@ def isfinite(self):
         """
         Returns True if the DFA accepts a finite language, False otherwise.
         """
+        try:
+            return self.maximum_word_length() != float('inf')
+        except ValueError:
+            return True
+
+    def count_words_of_length(self, k):
+        """
+        Counts words of length `k` accepted by the DFA
+        """
+        self._ensure_count_for_length(k)
+        return self._count_cache[k][self.initial_state]
+
+    def _ensure_count_for_length(self, k):
+        """
+        Populate count cache up to length k
+        """
+        while len(self._count_cache) <= k:
+            i = len(self._count_cache)
+            self._count_cache.append(defaultdict(int))
+            level = self._count_cache[i]
+            if i == 0:
+                level.update({state: 1 for state in self.final_states})
+            else:
+                prev_level = self._count_cache[i-1]
+                level.update({
+                    state: sum(prev_level[suffix_state] for suffix_state in self.transitions[state].values())
+                    for state in self.states
+                })
+
+    def words_of_length(self, k):
+        """
+        Generates all words of size k in the language represented by the DFA
+        """
+        self._ensure_words_of_length(k)
+        for word in self._word_cache[k][self.initial_state]:
+            yield word
+
+    def _ensure_words_of_length(self, k):
+        """
+        Populate word cache up to length k
+        """
+        sorted_symbols = sorted(self.input_symbols)
+        while len(self._word_cache) <= k:
+            i = len(self._word_cache)
+            self._word_cache.append(defaultdict(list))
+            level = self._word_cache[i]
+            if i == 0:
+                level.update({state: {''} for state in self.final_states})
```

Comment:
@Tagl (cc @eliotwrobson) If you updated the `defaultdict` to use a `list` instead of a `set`, shouldn't this dict comprehension also follow suit?

```python
level.update({state: [''] for state in self.final_states})
```

Reply:
Yes

Reply:
Updated